feat: Calendar Cache

volnei opened the pull request:
## 🎯 Overview

This PR introduces a comprehensive **Calendar Cache and Synchronization** system that significantly improves calendar performance and reliability by implementing local caching and real-time sync capabilities for external calendar integrations.

## How it works
<img width="1236" height="833" alt="image" src="https://github.com/user-attachments/assets/2d9ea56a-5600-451d-84c7-844865e4a5c3" />

## ✨ Key Features

### 📦 Calendar Event Caching
- **Local Storage**: Cache calendar events in the database for faster retrieval
- **Transparent Integration**: Existing calendar operations work unchanged via `CalendarCacheWrapper`
- **Selective Caching**: Only cache events marked as "busy" to optimize storage

### 🔄 Real-time Synchronization
- **Webhook Support**: Subscribe to calendar changes via provider webhooks
- **Provider Adapters**: Support for Google Calendar and Office365 calendar subscriptions
- **Sync Management**: Track sync status, tokens, and error handling

### 🏗️ Architecture

#### New Database Schema
```sql
-- Calendar cache events table
model CalendarCacheEvent {
  id                 String  @id @default(uuid())
  selectedCalendarId String
  externalId         String
  summary            String?
  description        String?
  location           String?
  start              DateTime
  end                DateTime
  isAllDay           Boolean
  timeZone           String?
  status             CalendarCacheEventStatus
  // ... additional fields
}

-- Enhanced SelectedCalendar with sync fields
model SelectedCalendar {
  // ... existing fields
  channelId          String?
  channelExpiration  DateTime?
  syncSubscribedAt   DateTime?
  syncToken          String?
  syncedAt           DateTime?
  syncErrorAt        DateTime?
  syncErrorCount     Int?
  calendarCacheEvents CalendarCacheEvent[]
}
```

#### Service Layer
- **`CalendarSubscriptionService`**: Manages calendar subscriptions and webhook handling
- **`CalendarCacheEventService`**: Handles event caching operations
- **`CalendarSyncService`**: Manages synchronization logic
- **`CalendarCacheWrapper`**: Transparent caching layer for existing Calendar interface

#### Provider Adapters
- **`GoogleCalendarSubscriptionAdapter`**: Google Calendar API integration
- **`Office365CalendarSubscriptionAdapter`**: Microsoft Graph API integration
- **`AdaptersFactory`**: Factory pattern for provider selection

## 🚀 Implementation Details

### Feature Flags
```typescript
"calendar-subscription-cache": boolean;  // Enable/disable caching
"calendar-subscription-sync": boolean;   // Enable/disable sync
```

### Cache Integration
The system integrates transparently with existing calendar operations:

```typescript
// Enhanced getCalendar function
if (CalendarCacheEventService.isAppSupported(credential.appId)) {
  const originalCalendar = new CalendarService(credential);
  return new CalendarCacheWrapper({
    originalCalendar,
    calendarCacheEventRepository
  });
}
```

### Webhook Handling
- **Google Calendar**: Uses `X-Goog-Channel-Token` and `X-Goog-Channel-ID` headers
- **Office365**: Uses Microsoft Graph subscription notifications
- **Validation**: Token-based webhook validation for security

## 🔧 Configuration

### Environment Variables
- `GOOGLE_WEBHOOK_TOKEN`: Token for validating Google Calendar webhooks
- `GOOGLE_WEBHOOK_URL`: Endpoint URL for Google Calendar notifications

### Repository Pattern
- **`SelectedCalendarRepository`**: Data access for calendar selections
- **`CalendarCacheEventRepository`**: Data access for cached events
- Clean separation of concerns with interface-based design

## 📊 Benefits

1. **Performance**: Faster calendar event retrieval from local cache
2. **Reliability**: Reduced dependency on external API availability
3. **Scalability**: Webhook-based sync reduces API polling overhead
4. **Flexibility**: Feature flags allow gradual rollout and testing
5. **Maintainability**: Clean architecture with clear separation of concerns

## 🚧 Current Status (WIP)

This is a work-in-progress implementation that includes:
- ✅ Core architecture and service layer
- ✅ Database schema and models
- ✅ Provider adapters (Google Calendar, Office365)
- ✅ Cache wrapper and repository pattern
- ✅ Feature flag integration
- 🔄 Webhook endpoint implementation (in progress)
- 🔄 Error handling and retry logic (in progress)
- 🔄 Performance testing and optimization (pending)

## 🧪 Testing

The implementation follows the existing Cal.com patterns and should be thoroughly tested with:
- Unit tests for service classes
- Integration tests for provider adapters  
- E2E tests for webhook handling
- Performance tests for cache effectiveness

---

This PR lays the foundation for a robust calendar caching and synchronization system that will significantly improve user experience and system performance for calendar operations.

volnei committed:
feat: calendar cache and sync - wip

volnei committed:
Add env.example

volnei committed:
refactor on CalendarCacheEventService

volnei committed:
remove test console.log

volnei committed:
Fix type checks errors

volnei committed:
chore: remove pt comment

volnei committed:
add route.ts

volnei committed:
chore: fix tests

volnei committed:
Improve cache impl

volnei committed:
chore: update recurring event id

volnei committed:
chore: small improvements

ThyMinimalDev commented on packages/features/calendar-subscription/lib/CalendarSubscriptionService.ts:
```diff
@@ -0,0 +1,225 @@
+
+    // cleanup cache after unsubscribe
+    if (await this.isCacheEnabled()) {
+      const { CalendarCacheEventService } = await import("./cache/CalendarCacheEventService");
+      const calendarCacheEventService = new CalendarCacheEventService(this.deps);
```
I see this dynamic import done multiple time within the class, should we either

- handle this with conditional DI
- have a private method to instantiate the service once:
```ts
private calendarCacheEventService: CalendarCacheEventService | null = null;

    private async getCacheService(): Promise<CalendarCacheEventService> {

        if (!this.calendarCacheEventService) {
            const { CalendarCacheEventService } = await import("./cache/CalendarCacheEventService");
            this.calendarCacheEventService = new CalendarCacheEventService(this.deps);
        }
        return this.calendarCacheEventService;
    }```

volnei replied:
Thank you @ThyMinimalDev, just moved to DI

volnei committed:
calendar cache improvements

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

volnei committed:
Fix remove dynamic imports

volnei committed:
Merge branch 'feat/calendar_sync_cache' of https://github.com/calcom/cal.com into feat/calendar_sync_cache

volnei committed:
Add cleanup stale cache

volnei committed:
Fix tests

volnei committed:
add event update

volnei commented:
@coderabbit summary

volnei committed:
type fixes

devin-ai-integration[bot] committed:
feat: add comprehensive tests for new calendar subscription API routes

- Add tests for /api/cron/calendar-subscriptions-cleanup route (9 tests)
- Add tests for /api/cron/calendar-subscriptions route (10 tests)
- Add tests for /api/webhooks/calendar-subscription/[provider] route (11 tests)
- Add missing feature flags for calendar-subscription-cache and calendar-subscription-sync
- All 30 tests pass with comprehensive coverage of authentication, feature flags, error handling, and service instantiation

Tests cover:
- Authentication scenarios (API key validation, Bearer tokens, query parameters)
- Feature flag combinations (cache/sync enabled/disabled states)
- Success and error handling (including non-Error exceptions)
- Service instantiation with proper dependency injection
- Provider validation for webhook endpoints

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
feat: add comprehensive tests for calendar subscription services, repositories, and adapters

- Add unit tests for CalendarSubscriptionService with subscription, webhook, and event processing
- Add unit tests for CalendarCacheEventService with cache operations and cleanup
- Add unit tests for CalendarSyncService with Cal.com event filtering and booking operations
- Add unit tests for CalendarCacheEventRepository with CRUD operations
- Add unit tests for SelectedCalendarRepository with calendar selection management
- Add unit tests for GoogleCalendarSubscriptionAdapter with subscription and event fetching
- Add unit tests for Office365CalendarSubscriptionAdapter with placeholder implementation
- Add unit tests for AdaptersFactory with provider management and adapter creation
- Fix lint issues by removing explicit 'any' type casting and unused variables
- All tests follow Cal.com conventions using Vitest framework with proper mocking

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: improve calendar-subscriptions-cleanup test performance by adding missing mocks

- Add comprehensive mocks for defaultResponderForAppDir, logger, performance monitoring, and Sentry
- Fix slow test execution (933ms -> <100ms) caused by missing dependency mocks
- Ensure consistent test performance across different environments

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Fix tests

volnei committed:
Fix tests

volnei committed:
type fix

volnei committed:
Fix coderabbit comments

volnei committed:
Fix types

volnei committed:
Fix test

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

emrysal reviewed (dismissed):
Nice work! I left a few comments, just some feedback we can evaluate before merge (improve logging, reduce verbosity keeping in mind we can tell a lot already in general, status codes and response codes don't require info's e.g.

Some other comments, nothing critical, this looks great!

volnei committed:
Update apps/web/app/api/cron/calendar-subscriptions/route.ts

Co-authored-by: Alex van Andel <me@alexvanandel.com>

volnei committed:
Fixes by first review

volnei committed:
merge conflict

devin-ai-integration[bot] committed:
feat: add database migrations for calendar cache and sync fields

- Add CalendarCacheEventStatus enum with confirmed, tentative, cancelled values
- Add new fields to SelectedCalendar: channelId, channelKind, channelResourceId, channelResourceUri, channelExpiration, syncSubscribedAt, syncToken, syncedAt, syncErrorAt, syncErrorCount
- Create CalendarCacheEvent table with foreign key to SelectedCalendar
- Add necessary indexes and constraints for performance and data integrity

Fixes database schema issues causing e2e test failures with 'column does not exist' errors.

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
only google-calendar for now

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

devin-ai-integration[bot] committed:
docs: add Calendar Cache and Sync feature documentation

- Add comprehensive feature overview and motivation
- Document feature flags with SQL examples
- Include SQL examples for enabling features for users and teams
- Reference technical documentation files

Addresses PR #23876 documentation requirements

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
docs: update calendar subscription README with comprehensive documentation

- Undo incorrect changes to main README.md
- Update packages/features/calendar-subscription/README.md with:
  - Feature overview and motivation
  - Environment variables section
  - Complete feature flags documentation with SQL examples
  - SQL examples for enabling features for users and teams
  - Detailed architecture documentation

Addresses PR #23876 documentation requirements

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
fix docs

keithwillcode committed:
Merge branch 'main' into feat/calendar_sync_cache

volnei committed:
Fix test to available calendars

volnei committed:
Fix test to available calendars

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

volnei committed:
add migration and sync boilerplate

volnei committed:
fix typo

volnei committed:
remove double log

volnei committed:
sync boilerplate

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

volnei committed:
remove console.log

volnei committed:
only subscribe for google calendar

volnei committed:
adjust for 3 months fetch

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

volnei committed:
only subscribe for teams that have feature enabled

volnei committed:
adjust tests

volnei committed:
chore: safe increment error count

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

volnei committed:
Merge branch 'main' into feat/calendar_sync_cache

volnei committed:
Fix test

volnei committed:
Merge branch 'feat/calendar_sync_cache' of https://github.com/calcom/cal.com into feat/calendar_sync_cache

emrysal committed:
Rename findAllBySelectedCalendarIds to findAllBySelectedCalendarIdsbetween + minor nits

volnei committed:
Fix tests

emrysal committed:
Add more declaritive code for increased clarity

emrysal reviewed (approved):
🥼 - Let's get this out there! 🔥
